src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -29,10 +29,27 @@ public abstract class AbstractLanguageServer<TRequestContext> : ILifeCycleManage
 
     public bool IsInitialized { get; private set; }
 
-    // Fields used during shutdown.
-    private bool _shuttingDown;
+    /// <summary>
+    /// Ensures that we only run shutdown and exit code once in order.
+    /// Guards access to <see cref="_shutdownRequestTask"/> and <see cref="_exitNotificationTask"/>
+    /// </summary>
+    private readonly object _lifeCycleLock = new();
+
+    /// <summary>
+    /// Task representing the work done on LSP server shutdown.
+    /// </summary>
+    private Task? _shutdownRequestTask;
 
-    public bool HasShutdownStarted => _shuttingDown;
+    /// <summary>
+    /// Task representing the work down on LSP exit.
+    /// </summary>
+    private Task? _exitNotificationTask;
+
+    /// <summary>
+    /// Task completion source that is started when the server starts and completes when the server exits.
+    /// Used when callers need to wait for the server to cleanup.
+    /// </summary>
+    private readonly TaskCompletionSource<bool> _serverExitedSource = new(false);
```

Comment:
what does the 'false' mean here?

Reply:
named parameter?

Reply:
I just needed a type, but bool could be misleading.  Changing to `object?`